src/NHibernate/Linq/DefaultQueryProvider.cs

```diff
@@ -59,12 +60,32 @@ public virtual IQueryable<T> CreateQuery<T>(Expression expression)
 			return new NhQueryable<T>(this, expression);
 		}
 
-		public virtual object ExecuteFuture(Expression expression)
+		public virtual IEnumerable<TResult> ExecuteFuture<TResult>(Expression expression)
 		{
-			IQuery query;
-			NhLinqExpression nhQuery;
-			NhLinqExpression nhLinqExpression = PrepareQuery(expression, out query, out nhQuery);
-			return ExecuteFutureQuery(nhLinqExpression, query, nhQuery);
+			PrepareQuery(expression, out var query, out var nhQuery);
+
+			var result = query.Future<TResult>();
+			SetupFutureResult(nhQuery, (IDelayedValue)result);
```

Comment:
Previously the cast was done only if a post execute transformer was defined, hiding the potential failure of the lack of multi-query fallback. There were not actual bug currently because there is up to now no way to get a post result transformer with something else than a future value, and `FutureValue` own fallback was already yielding a `IDelayedValue`.